client.go

```diff
@@ -54,48 +46,106 @@ func (c *ExtensionManagerClient) Close() {
 func (c *ExtensionManagerClient) RegisterExtension(info *osquery.InternalExtensionInfo, registry osquery.ExtensionRegistry) (*osquery.ExtensionStatus, error) {
-	return c.Client.RegisterExtension(context.Background(), info, registry)
+	return c.RegisterExtensionContext(context.Background(), info, registry)
 }
 
-// DeregisterExtension de-registers the extension plugins with the osquery process.
+// RegisterExtension registers the extension plugins with the osquery process.
+func (c *ExtensionManagerClient) RegisterExtensionContext(ctx context.Context, info *osquery.InternalExtensionInfo, registry osquery.ExtensionRegistry) (*osquery.ExtensionStatus, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.Client.RegisterExtension(ctx, info, registry)
+}
+
+// DeregisterExtension de-registers the extension plugins with the osquery process, using a new background context
 func (c *ExtensionManagerClient) DeregisterExtension(uuid osquery.ExtensionRouteUUID) (*osquery.ExtensionStatus, error) {
-	return c.Client.DeregisterExtension(context.Background(), uuid)
+	return c.DeregisterExtensionContext(context.Background(), uuid)
 }
 
-// Options requests the list of bootstrap or configuration options.
+// DeregisterExtension de-registers the extension plugins with the osquery process.
+func (c *ExtensionManagerClient) DeregisterExtensionContext(ctx context.Context, uuid osquery.ExtensionRouteUUID) (*osquery.ExtensionStatus, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.Client.DeregisterExtension(ctx, uuid)
+}
+
+// Options requests the list of bootstrap or configuration options, using a new background context.
 func (c *ExtensionManagerClient) Options() (osquery.InternalOptionList, error) {
-	return c.Client.Options(context.Background())
+	return c.OptionsContext(context.Background())
+}
+
+// Options requests the list of bootstrap or configuration options.
+func (c *ExtensionManagerClient) OptionsContext(ctx context.Context) (osquery.InternalOptionList, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.Client.Options(ctx)
+}
+
+// Query requests a query to be run and returns the extension
+// response, using a new background context.  Consider using the
+// QueryRow or QueryRows helpers for a more friendly interface.
+func (c *ExtensionManagerClient) Query(sql string) (*osquery.ExtensionResponse, error) {
+	return c.QueryContext(context.Background(), sql)
 }
 
 // Query requests a query to be run and returns the extension response.
 // Consider using the QueryRow or QueryRows helpers for a more friendly
 // interface.
-func (c *ExtensionManagerClient) Query(sql string) (*osquery.ExtensionResponse, error) {
-	return c.Client.Query(context.Background(), sql)
+func (c *ExtensionManagerClient) QueryContext(ctx context.Context, sql string) (*osquery.ExtensionResponse, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.Client.Query(ctx, sql)
```

Comment:
We had to use ```c.Client.Query``` directly because we wanted to pass cancellable context and the c.Client is public. After this change API is still public, the calls are still not serialized and can be called concurrently.

Reply:
Which part isn't serialized here?

I'm thinking about whether it makes sense to drop the mutexes in favor of a rate.Limiter curious what you think

Reply:
> Which part isn't serialized here?

In our code we used ```c.Client.Query``` directly because there were no API with context exposed on ```ExtensionManagerClient```. With this PR we can switch to ```QueryContext``` and drop our own locking/synchronization.

Reply:
> I'm thinking about whether it makes sense to drop the mutexes in favor of a rate.Limiter curious what you think

What property of rate.Limiter would be appealing in this case? 
As far as I understand the protocol breaks if there is more than one request in flight. Mutex seems appropriate.

Reply:
Ok, this makes sense: time out the request at the ```limiter.Wait``` if the context.WithTimeout is used, instead of timing out when the previous request is completed and you get the change to grab the mutex lock.

Also, thinking of the cases where we need to limit the query running time: for example, do not run query for longer than a minute. Currently this is only possible with setting the deadline on the context of the query (context.WithTimeout).
In both cases the context timeout would not be exactly the query running time limit though. 
One bad (long running) query can cause the pending queries contexts to timeout. By the time the query gets to run, the context might be already in the ```context deadline exceeded``` state. 
Thoughts?

Another question: if we time out through the context, does this cancel the query at the protocol level and the server, meaning the guarantee that no response will be sent back for the timed out query, when we already sent the next query request (not familiar with that part)?

Reply:
It's been a year, but I've resurrected this over on #108. I think I've found a better idiom for handling timeouts and context cancelation.

I'm not sure what the underlying thrift protocol does if you cancel the context for a running query, I'm not sure this SDK can do much